book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+update from its root to its tip.
+
+The `Chain` type is used to represent the non-finalized portion of a complete
+chain of blocks rooted at the genesis block. The parent block of the root of
+a `Chain` is the tip of the finalized portion of the chain.
+
+The `Chain` type supports serveral operations to manipulate chains, `push`,
+`pop_root`, and `fork`. `push` is the most fundamental operation and handles
+contextual validation of chains as they are extended. `pop_root` is provided
+for finalization, and is how we move blocks from the non-finalized portion of
+the state to the finalized portion. `fork` on the other hand handles creating
+new chains for `push` when new blocks arrive whose parent isn't a tip of an
+existing chain.
+
+The `Chain` type is defined by the following struct and API:
+
+```rust
+struct Chain {
+    blocks: BTreeMap<block::Height, Arc<Block>>,
+    height_by_hash: HashMap<block::Hash, block::Height>,
+    tx_by_hash: HashMap<transaction::Hash, (block::Height, tx_index)>,
+
+    utxos: HashSet<transparent::Output>,
+    sapling_anchors: HashSet<sapling::tree::Root>,
+    sprout_anchors: HashSet<sprout::tree::Root>,
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+    partial_cumulative_work: PartialCumulativeWork,
+}
+```
+
+#### `pub fn push(&mut self, block: Arc<Block>) -> Result<(), Error>`
+
+Push a block into a chain as the new tip if the block is a valid extension of
+that chain.
+
+1. Run contextual validation checks on block against Self
+
+1. Update cumulative data members
+    - Add block to end of `self.blocks`
+    - Add hash to `height_by_hash`
+    - for each `transaction` in `block`
+      - add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+    - Add new utxos and remove consumed utxos from `self.utxos`
+    - Add anchors to the appropriate `self.<version>_anchors`
+    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add work to `self.partial_cumulative_work`
+
+#### `pub fn pop_root(&mut self) -> Arc<Block>`
+
+Remove the lowest height block of the non-finalized portion of a chain.
+
+1. Remove the lowest height block from `self.blocks`
+
+1. Update cumulative data members
+    - Remove the block's hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Remove new utxos from `self.utxos`
+    - Remove the anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+
+1. Return the block
+
+**Note**: We do not subtract work from `self.partial_cummulative_work`. This
+is to make sure that the ordering of chains is stable while finalizing blocks.
+
+#### `pub fn fork(&self, new_tip: block::Hash) -> Option<Self>`
+
+Fork a chain at the block with the given hash, if it is part of this chain.
+
+1. If `self` does not contain `new_tip` return `None`
+
```

Comment:
```suggestion
1. If `new_tip` is equal to the finalized tip, return an empty Chain
  - see `push: initialise cumulative data members` for details
```

Reply:
I don't think this can or should be handled by the `Chain` structure but this raises a good point. The `Chain` and `ChainSet` both have no notion of the finalized state, so the 3rd check that I've been proposing cant be done with the context currently available when commiting these blocks to the in memory state.

I've only just started thinking about this but my initial instinct would be to have this check be done before calling `chain_set.queue(block)`, as part of the ## Committing non-finalized blocks section. Instead it would be `1.` compare the parent hash against the current tip of the finalized chain, then we'd need some way to insert it, could add a new method for inserting new chains directly, not in love with this idea yet, so I'll keep thinking.

Reply:
actually, this probably means that we just need to give the `Chain` and `ChainSet` structures access to the finalized state, I imagine they'll need it for the contextual verification checks.

Reply:
Then we need to update the `ChainSet` spec and implementation, so it handles adding the first `Chain` to an empty `ChainSet`

Reply:
Turns out this already happened at https://github.com/ZcashFoundation/zebra/pull/902/files#diff-0a5b73420cea8c2ce73f46bc2b5d56c7R495